File: examples/experiments/multilang/python/graph_pagerank.py
from array import array


def _int_param(params, name, default, minimum, maximum):
    try:
        value = int(params.get(name, default))
    except (TypeError, ValueError, AttributeError):
        value = default
    return max(minimum, min(maximum, value))


def _build_graph(nodes, degree):
    edges = array("I", [0]) * (nodes * degree)

    for u in range(nodes):
        base = u * degree
        state = (u * 2654435761 + 1013904223) & 0xFFFFFFFF

        for j in range(degree):
            state = (1664525 * state + 1013904223 + j) & 0xFFFFFFFF
            v = state % nodes

            if v == u:
                v = (v + j + 1) % nodes

            edges[base + j] = v

    return edges


def _pagerank(nodes, degree, edges, iterations, damping):
    ranks = [1.0 / nodes] * nodes
    base_rank = (1.0 - damping) / nodes

    for _ in range(iterations):
        new_ranks = [base_rank] * nodes

        for u in range(nodes):
            contribution = damping * ranks[u] / degree
            offset = u * degree

            for j in range(degree):
                new_ranks[edges[offset + j]] += contribution

        ranks = new_ranks

    return ranks


def handler(params, context):
    params = params or {}

    nodes = _int_param(params, "nodes", 100_000, 5_000, 200_000)
    degree = _int_param(params, "degree", 8, 2, 16)
    iterations = _int_param(params, "iterations", 50, 5, 80)
    damping = 0.85

    edges = _build_graph(nodes, degree)
    ranks = _pagerank(nodes, degree, edges, iterations, damping)

    rank_sum = sum(ranks)
    sample_sum = 0.0
    step = max(1, nodes // 1000)

    for i in range(0, nodes, step):
        sample_sum += ranks[i]

    return {
        "benchmark": "graph_pagerank",
        "nodes": nodes,
        "degree": degree,
        "iterations": iterations,
        "rank0": round(ranks[0], 12),
        "rank_sum": round(rank_sum, 12),
        "sample_sum": round(sample_sum, 12),
    }

File: examples/experiments/multilang/nodejs/graph_pagerank.js
function intParam(params, name, defaultValue, min, max) {
  const parsed = Number.parseInt(params?.[name] ?? defaultValue, 10);
  const value = Number.isFinite(parsed) ? parsed : defaultValue;
  return Math.max(min, Math.min(max, value));
}

function buildGraph(nodes, degree) {
  const edges = new Uint32Array(nodes * degree);
  for (let u = 0; u < nodes; u++) {
    const base = u * degree;
    let state = (Math.imul(u, 0x9e3779b1) + 1013904223) >>> 0;
    for (let j = 0; j < degree; j++) {
      state = (Math.imul(1664525, state) + 1013904223 + j) >>> 0;
      let v = state % nodes;
      if (v === u) {
        v = (v + j + 1) % nodes;
      }
      edges[base + j] = v;
    }
  }
  return edges;
}

function pagerank(nodes, degree, edges, iterations, damping) {
  let ranks = new Float64Array(nodes);
  ranks.fill(1.0 / nodes);
  const baseRank = (1.0 - damping) / nodes;
  for (let iter = 0; iter < iterations; iter++) {
    const newRanks = new Float64Array(nodes);
    newRanks.fill(baseRank);
    for (let u = 0; u < nodes; u++) {
      const contribution = (damping * ranks[u]) / degree;
      const offset = u * degree;
      for (let j = 0; j < degree; j++) {
        newRanks[edges[offset + j]] += contribution;
      }
    }
    ranks = newRanks;
  }
  return ranks;
}

module.exports = (params, ctx) => {
  params = params || {};
  const nodes = intParam(params, "nodes", 100000, 5000, 200000);
  const degree = intParam(params, "degree", 8, 2, 16);
  const iterations = intParam(params, "iterations", 50, 5, 80);
  const damping = 0.85;
  const edges = buildGraph(nodes, degree);
  const ranks = pagerank(nodes, degree, edges, iterations, damping);
  let rankSum = 0.0;
  for (let i = 0; i < nodes; i++) {
    rankSum += ranks[i];
  }
  let sampleSum = 0.0;
  const step = Math.max(1, Math.floor(nodes / 1000));
  for (let i = 0; i < nodes; i += step) {
    sampleSum += ranks[i];
  }
  return {
    benchmark: "graph_pagerank",
    nodes,
    degree,
    iterations,
    rank0: Number(ranks[0].toFixed(12)),
    rank_sum: Number(rankSum.toFixed(12)),
    sample_sum: Number(sampleSum.toFixed(12)),
  };
};
